# Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto

<b> Rahul Nayak<b>

## PART 1

### Importing data from wiki page.

In [14]:
import requests
import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = requests.get(url).content
df_list = pd.read_html(html)
df = df_list[0]
print(df)

    Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                                  NaN  
1                                                  NaN  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                                NaN  
176                                                NaN  
177                                       

In [16]:
column_names = ['Postal Code', 'Borough', 'Neighborhood'] 
raw_data = pd.DataFrame(columns=column_names)
raw_data=df

In [20]:
raw_data.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

### Cleaning the raw data

First we will drop rows where bourough is 'Not assigned'

In [56]:
logi=raw_data['Borough']=="Not assigned"
raw_data=raw_data[~logi]
raw_data.head(10)

Postal Code           Borough                                 Neighborhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke                             Islington Avenue
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills
12         M4B         East York              Parkview Hill, Woodbine Gardens
13         M5B  Downtown Toronto                     Garden District, Ryerson

In [33]:
raw_data.shape

(103, 3)

We find that earlier there were 180 rows and now there are 103 rows i.e 77 rows had Borough not assigned.

Now we will replace the 'Not assigned' in the Neighborhood colomn with the Borough name. 

In [45]:
logi2=raw_data['Neighborhood']=="Not assigned"
logi21=raw_data['Neighborhood']==""

In [44]:
logi2.value_counts()

False    103
Name: Neighborhood, dtype: int64

In [46]:
logi21.value_counts()

False    103
Name: Neighborhood, dtype: int64

We find that all the neighborhoods are labelled and we can proceed

now we will find neighborhoods with the same postal code.

In [39]:
logi3=raw_data['Postal Code']
print(logi3.value_counts())

M1L    1
M3L    1
M1E    1
M7R    1
M1G    1
      ..
M6R    1
M9W    1
M4Y    1
M1R    1
M5K    1
Name: Postal Code, Length: 103, dtype: int64


In [53]:
logi4=logi3.value_counts()>1
logi4.value_counts()

False    103
Name: Postal Code, dtype: int64

We find that all the postal codes apper only once i.e if there are more than one neighborhood for a postal code, it has already been merged into one row.

In [59]:
clean_data=raw_data

In [60]:
clean_data.shape

(103, 3)

<b> We have our cleaned data with 103 rows and 3 colomns.<b>

____________________________________________________________________________________________________________________________

## Part 2

### Getting latitude and longitude values

We will use the csv files that contain the latitude and longitude data.

In [61]:
loc_data = pd.read_csv('https://cocl.us/Geospatial_data')

In [62]:
loc_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

we will have to append the colomns from here to the clean_data, for which we will sort both the dataframes acoording to Postal code in ascending order.

In [71]:
clean_data=clean_data.sort_values("Postal Code", axis=0, ascending=True)

To reset the indexing

In [257]:
clean_data.to_csv('test1.csv', index=False)

In [258]:
clean_data=pd.read_csv('test1.csv')

Now we will append the lat and long values to clean_data

In [259]:
clean_data['Latitude']=loc_data['Latitude']
clean_data['Longitude']=loc_data['Longitude']

In [260]:
clean_data.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

But here there are a lot of NaN values in Latitude and longitude values, we will have to remove them.

In [261]:
clean_data.dropna(subset=["Latitude","Longitude"],inplace=True)

_________________________

## Part 3

### We will Explore and cluster the neighborhoods in Toronto

Let's get the geographical coordinates of Toronto.

First we will import required libraries

In [78]:
!conda install -c conda-forge geopy --yes

Solving environment: ...working... 
  - anaconda/win-64::ca-certificates-2020.1.1-0, anaconda/win-64::openssl-1.1.1d-he774522_4
  - anaconda/win-64::ca-certificates-2020.1.1-0, defaults/win-64::openssl-1.1.1d-he774522_4
  - anaconda/win-64::openssl-1.1.1d-he774522_4, defaults/win-64::ca-certificates-2020.1.1-0
  - defaults/win-64::ca-certificates-2020.1.1-0, defaults/win-64::openssl-1.1.1d-he774522_4done

## Package Plan ##

  environment location: C:\Users\rahul\anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

  ge

In [79]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Use geopy library to get the latitude and longitude values of Toronto City.

In [262]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Let us visualise Toronto with its neighbourhood.

In [263]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

In [264]:
# add markers to map
for lat, lng, label in zip(clean_data['Latitude'], clean_data['Longitude'], clean_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

Define Foursquare Credentials and Version

In [314]:
# @hidden_cell
CLIENT_ID = 'BY2T4MPZF3K3GVRJ0GTE1PHOSX21JTPMWRIGYOBYG4WJQBTN' # your Foursquare ID
CLIENT_SECRET = '0SL3P3YYCVSJ3UEBNXBODCMSO4DUA15EAET1JRYHZLPM2D1H' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: BY2T4MPZF3K3GVRJ0GTE1PHOSX21JTPMWRIGYOBYG4WJQBTN
CLIENT_SECRET:0SL3P3YYCVSJ3UEBNXBODCMSO4DUA15EAET1JRYHZLPM2D1H


We will explore a neighbourhood at random

In [266]:
clean_data.loc[36, 'Neighborhood']

'Studio District'

Get the neighborhood's latitude and longitude values.

In [267]:
neighborhood_latitude = clean_data.loc[36, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = clean_data.loc[36, 'Longitude'] # neighborhood longitude value

neighborhood_name = clean_data.loc[36, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Studio District are 43.695343900000005, -79.3183887.


Now, let's get the top 100 venues that are in Studio District within a radius of 500 meters.

First, let's create the GET request URL.

In [268]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=BY2T4MPZF3K3GVRJ0GTE1PHOSX21JTPMWRIGYOBYG4WJQBTN&client_secret=0SL3P3YYCVSJ3UEBNXBODCMSO4DUA15EAET1JRYHZLPM2D1H&v=20180605&ll=43.695343900000005,-79.3183887&radius=500&limit=100'

Send the GET request and examine the resutls


In [269]:
results = requests.get(url).json()

In [270]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [271]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

C:\Users\rahul\anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name      categories        lat        lng
0               East York Memorial Arena    Skating Rink  43.697224 -79.315397
1                 East York Curling Club     Curling Ice  43.696827 -79.313658
2                         The Beer Store      Beer Store  43.693731 -79.316759
3                     Shoppers Drug Mart        Pharmacy  43.699244 -79.318939
4  Beauty Boutique by Shoppers Drug Mart  Cosmetics Shop  43.699232 -79.319319

Let's create a function to repeat the same process to all the neighborhoods in toronto

In [272]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now we will write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [273]:
toronto_venues = getNearbyVenues(names=clean_data['Neighborhood'],
                                   latitudes=clean_data['Latitude'],
                                   longitudes=clean_data['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale
York Mills West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, T

Let's check the size of the resulting dataframe

In [274]:
print(toronto_venues.shape)
toronto_venues.head()

(1570, 7)


Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1                          Malvern, Rouge              43.806686   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3  Rouge Hill, Port Union, Highland Creek              43.784535   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                      Venue  Venue Latitude  \
0              -79.194353                    Wendy’s       43.807448   
1              -79.194353      Interprovincial Group       43.805630   
2              -79.160497      Royal Canadian Legion       43.782533   
3              -79.160497  Affordable Toronto Movers       43.787919   
4              -79.188711             RBC Royal Bank       43.766790   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.200378            Print Shop  
2       -79.163085                   Bar  
3       -79.162977         Moving Target  
4       -79.191151                  Bank

Let's check how many venues were returned for each neighborhood

In [275]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Bayview Village                                                        14   
Bedford Park, Lawrence Manor East                                       3   
Berczy Park                                                            19   
Birch Cliff, Cliffside West                                             5   
Brockton, Parkdale Village, Exhibition Place                           65   
Caledonia-Fairbanks                                                    46   
Cedarbrae                                                               8   
Central Bay Street                                                     40   
Christie                                                              100   
Cliffside, Cliffcrest, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                         38   
Davisville North                                                       33   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                      100   
Don Mills                                                               4   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview                                                              16   
Dufferin, Dovercourt Village                                           77   
East Toronto                                                            4   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      4   
Fairview, Henry Farm, Oriole                                            4   
Garden District, Ryerson                                                4   
Glencairn                                                              45   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                       7   
Hillcrest Village                                                       4   
Humber Summit                                                          24   
Humberlea, Emery                                                       36   
Humewood-Cedarvale                                                     77   
India Bazaar, The Beaches West                                         10   
Islington Avenue                                                       24   
Kennedy Park, Ionview, East Birchmount Park                             8   
Lawrence Manor, Lawrence Heights                                        4   
Lawrence Park                                                           4   
Leaside                                                                 6   
Little Portugal, Trinity                                               54   
Malvern, Rouge                                                          2   
North Park, Maple Leaf Park, Upwood Park                               92   
Northwood Park, York University                                        32   
Parkview Hill, Woodbine Gardens                                        21   
Parkwoods                                                              67   
Queen's Park, Ontario Provincial Government                           100   
Regent Park, Harbourfront                                              22   
Richmond, Adelaide, King                                                3   
Roselawn                                                               16   
Rouge Hill, Port Union, Highland Creek                                  2   
Runnymede, The Junction North                                         100   
Scarborough Village                                                     2   
St. James Town                                                         42   
Studio District                                                         9   
The Beaches                                     

Let's find out how many unique categories can be curated from all the returned venues

In [276]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 238 uniques categories.


Now we will Analyze Each Neighborhood

In [283]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

In [284]:
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] =toronto_venues['Neighborhood'] 
toronto_onehot.head()

Afghan Restaurant  Airport  American Restaurant  Antique Shop  Aquarium  \
0                  0        0                    0             0         0   
1                  0        0                    0             0         0   
2                  0        0                    0             0         0   
3                  0        0                    0             0         0   
4                  0        0                    0             0         0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0       0     0    0   
1                   0          0           0       0     0    0   
2                   0          0           0       0     0    1   
3                   0          0           0       0     0    0   
4                   0          0           0       0     1    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0               0                 0                   0      0   
1               0                 0                   0      0   
2               0                 0                   0      0   
3               0                 0                   0      0   
4               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0       0          0         0                     0               0        0   
1       0          0         0                     0               0        0   
2       0          0         0                     0               0        0   
3       0          0         0                     0               0        0   
4       0          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0            0                0         0             0              0   
1            0                0         0             0              0   
2            0                0         0             0              0   
3            0                0         0             0              0   
4            0                0         0             0              0   

   Bus Line  Bus Station  Bus Stop  Business Service  Butcher  Café  \
0         0            0         0                 0        0     0   
1         0            0         0                 0        0     0   
2         0            0         0                 0        0     0   
3         0            0         0                 0        0     0   
4         0            0         0                 0        0     0   

   Camera Store  Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
0             0                     0            0                   0   
1             0                     0            0                   0   
2             0                     0            0                   0   
3             0                     0            0                   0   
4             0                     0            0                   0   

   Chocolate Shop  Clothing Store  Cocktail Bar  Coffee Shop  \


In [285]:
# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-74]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head(10)

Neighborhood  Afghan Restaurant  Airport  \
0                          Malvern, Rouge                  0        0   
1                          Malvern, Rouge                  0        0   
2  Rouge Hill, Port Union, Highland Creek                  0        0   
3  Rouge Hill, Port Union, Highland Creek                  0        0   
4       Guildwood, Morningside, West Hill                  0        0   
5       Guildwood, Morningside, West Hill                  0        0   
6       Guildwood, Morningside, West Hill                  0        0   
7       Guildwood, Morningside, West Hill                  0        0   
8       Guildwood, Morningside, West Hill                  0        0   
9       Guildwood, Morningside, West Hill                  0        0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   
5                    0             0         0            0           0   
6                    0             0         0            0           0   
7                    0             0         0            0           0   
8                    0             0         0            0           0   
9                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          0   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   
5                    0                 0                   0          0   
6                    0                 0                   0          0   
7                    0                 0                   0          0   
8                    0                 0                   0          0   
9                    0                 0                   0          0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    1               0                 0   
3           0       0     0    0               0                 0   
4           0       0     1    0               0                 0   
5           0       0     0    0               0                 0   
6           0       0     0    0               0                 0   
7           0       0     0    0               0                 0   
8           0       0     0    0               0                 0   
9           0       0     0    0               0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   
5                   0      0                0         0           0   
6                   0      0                0         0           0   
7                   0      0                0         0           0   
8                   0      0                0         0           0   
9                   0      0                0         0           0   

   Belgian Restauran

And let's examine the new dataframe size.

In [286]:
toronto_onehot.shape

(1570, 238)

In [287]:
toronto_onehot.to_csv('test.csv')

In [288]:
toronto_onehot=pd.read_csv('test.csv')

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [289]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped=toronto_grouped.drop(['Unnamed: 0'], axis=1)
toronto_grouped.head()

Neighborhood  Afghan Restaurant  Airport  \
0                               Bayview Village                0.0      0.0   
1             Bedford Park, Lawrence Manor East                0.0      0.0   
2                                   Berczy Park                0.0      0.0   
3                   Birch Cliff, Cliffside West                0.0      0.0   
4  Brockton, Parkdale Village, Exhibition Place                0.0      0.0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                  0.0           0.0       0.0          0.0    0.000000   
1                  0.0           0.0       0.0          0.0    0.000000   
2                  0.0           0.0       0.0          0.0    0.000000   
3                  0.0           0.0       0.0          0.0    0.000000   
4                  0.0           0.0       0.0          0.0    0.015385   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                  0.0               0.0                 0.0        0.0   
1                  0.0               0.0                 0.0        0.0   
2                  0.0               0.0                 0.0        0.0   
3                  0.0               0.0                 0.0        0.0   
4                  0.0               0.0                 0.0        0.0   

   Bagel Shop  Bakery      Bank       Bar  Baseball Field  Baseball Stadium  \
0         0.0     0.0  0.071429  0.000000             0.0               0.0   
1         0.0     0.0  0.000000  0.000000             0.0               0.0   
2         0.0     0.0  0.000000  0.000000             0.0               0.0   
3         0.0     0.0  0.000000  0.000000             0.0               0.0   
4         0.0     0.0  0.015385  0.030769             0.0               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0         0.0   
1                 0.0    0.0              0.0       0.0         0.0   
2                 0.0    0.0              0.0       0.0         0.0   
3                 0.0    0.0              0.0       0.0         0.0   
4                 0.0    0.0              0.0       0.0         0.0   

   Belgian Restaurant  Bike Shop  Bistro  Bookstore  Boutique  \
0                 0.0        0.0     0.0        0.0       0.0   
1                 0.0        0.0     0.0        0.0       0.0   
2                 0.0        0.0     0.0        0.0       0.0   
3                 0.0        0.0     0.0        0.0       0.0   
4                 0.0        0.0     0.0        0.0       0.0   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bridal Shop  \
0                   0.0             0.0  0.000000          0.0   
1                   0.0             0.0  0.000000          0.0   
2                   0.0             0.0  0.052632          0.0   
3                   0.0             0.0  0.000000          0.0   
4                   0.0             0.0  0.000000          0.0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0         0.000000       0.0      0.000000       0.000000       0.0   
1         0.000000       0.0      0.000000       0.000000       0.0   
2         0.000000       0.0      0.000000       0.052632       0.0   
3         0.000000       0.0      0.000000       0.000000       0.0   
4         0.030769       0.0      0.030769       0.000000       0.0   

   Bus Station  Bus Stop  Business Service  Butcher      Café  Camera Store  \
0          0.0       0.0          0.000000      0.0  0.000000           0.0   
1          0.0       0.0          0.000000      0.0  0.000000           0.0   
2          0.0       0.0          0.000000      0.0  0.000000           0.0   
3          0.0       0.0          0.000000      0.0  0.200000           0.0   
4          0.0       0.0          0.015385      0.0  0.046154           0.0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolat

Let's confirm the new size

In [290]:
toronto_grouped.shape

(62, 237)

Let's print each neighborhood along with the top 5 most common venues

In [291]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bayview Village----
             venue  freq
0      Pizza Place  0.14
1    Shopping Mall  0.07
2      Coffee Shop  0.07
3      Gas Station  0.07
4  Thai Restaurant  0.07


----Bedford Park, Lawrence Manor East----
               venue  freq
0               Park  0.33
1              Trail  0.33
2         Restaurant  0.33
3  Afghan Restaurant  0.00
4      Moving Target  0.00


----Berczy Park----
                  venue  freq
0  Fast Food Restaurant  0.11
1           Pizza Place  0.05
2         Movie Theater  0.05
3           Coffee Shop  0.05
4          Liquor Store  0.05


----Birch Cliff, Cliffside West----
                   venue  freq
0           Skating Rink   0.2
1        College Stadium   0.2
2                   Café   0.2
3                   Farm   0.2
4  General Entertainment   0.2


----Brockton, Parkdale Village, Exhibition Place----
                venue  freq
0         Coffee Shop  0.17
1      Sandwich Place  0.06
2  Italian Restaurant  0.06
3                Café  0.05

                  venue  freq
0         Metro Station  0.14
1        Massage Studio  0.14
2    Miscellaneous Shop  0.14
3    Falafel Restaurant  0.14
4  Caribbean Restaurant  0.14


----York Mills West----
                      venue  freq
0                   Dog Run  0.25
1                      Pool  0.25
2               Golf Course  0.25
3  Mediterranean Restaurant  0.25
4                 Pet Store  0.00


----York Mills, Silver Hills----
                      venue  freq
0                Playground   0.5
1                      Park   0.5
2                    Museum   0.0
3  Mediterranean Restaurant   0.0
4               Men's Store   0.0




Let's put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.

In [292]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [293]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                               Bayview Village           Pizza Place   
1             Bedford Park, Lawrence Manor East                  Park   
2                                   Berczy Park  Fast Food Restaurant   
3                   Birch Cliff, Cliffside West                  Farm   
4  Brockton, Parkdale Village, Exhibition Place           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0       Thai Restaurant              Pharmacy          Noodle House   
1                 Trail            Restaurant         Women's Store   
2             Pet Store           Pizza Place               Brewery   
3       College Stadium          Skating Rink                  Café   
4    Italian Restaurant        Sandwich Place                  Café   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Coffee Shop    Chinese Restaurant         Shopping Mall   
1    Distribution Center  Fast Food Restaurant        Farmers Market   
2          Burrito Place        Sandwich Place            Restaurant   
3  General Entertainment         Women's Store               Dog Run   
4           Burger Joint        Ice Cream Shop                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant   Fried Chicken Joint     Italian Restaurant  
1                  Farm    Falafel Restaurant            Event Space  
2                   Pub           Coffee Shop                   Park  
3  Fast Food Restaurant        Farmers Market     Falafel Restaurant  
4       Thai Restaurant   Japanese Restaurant        Bubble Tea Shop

Run k-means to cluster the neighborhood into 5 clusters.

In [294]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [295]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

neighborhoods_venues_sorted.head()

Cluster Labels                                  Neighborhood  \
0               1                               Bayview Village   
1               0             Bedford Park, Lawrence Manor East   
2               1                                   Berczy Park   
3               1                   Birch Cliff, Cliffside West   
4               1  Brockton, Parkdale Village, Exhibition Place   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Pizza Place       Thai Restaurant              Pharmacy   
1                  Park                 Trail            Restaurant   
2  Fast Food Restaurant             Pet Store           Pizza Place   
3                  Farm       College Stadium          Skating Rink   
4           Coffee Shop    Italian Restaurant        Sandwich Place   

  4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0          Noodle House            Coffee Shop    Chinese Restaurant   
1         Women's Store    Distribution Center  Fast Food Restaurant   
2               Brewery          Burrito Place        Sandwich Place   
3                  Café  General Entertainment         Women's Store   
4                  Café           Burger Joint        Ice Cream Shop   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0         Shopping Mall  Fast Food Restaurant   Fried Chicken Joint   
1        Farmers Market                  Farm    Falafel Restaurant   
2            Restaurant                   Pub           Coffee Shop   
3               Dog Run  Fast Food Restaurant        Farmers Market   
4                   Bar       Thai Restaurant   Japanese Restaurant   

  10th Most Common Venue  
0     Italian Restaurant  
1            Event Space  
2                   Park  
3     Falafel Restaurant  
4        Bubble Tea Shop

In [296]:
toronto_merged = clean_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             1.0  Fast Food Restaurant            Print Shop   
1 -79.160497             1.0                   Bar         Moving Target   
2 -79.188711             1.0    Mexican Restaurant     Electronics Store   
3 -79.216917             1.0           Coffee Shop     Korean Restaurant   
4 -79.239476             1.0   Fried Chicken Joint      Hakka Restaurant   

  3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
0         Women's Store         Distribution Center   Filipino Restaurant   
1               Dog Run  Financial or Legal Service   Filipino Restaurant   
2                  Bank               Moving Target   Rental Car Location   
3     Indian Restaurant               Women's Store      Doner Restaurant   
4  Caribbean Restaurant                        Bank                Bakery   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Farmers Market                  Farm    Falafel Restaurant   
1        Fast Food Restaurant        Farmers Market                  Farm   
2                Intersection        Medical Center        Breakfast Spot   
3  Financial or Legal Service   Filipino Restaurant  Fast Food Restaurant   
4             Thai Restaurant           Gas Station    Athletics & Sports   

  9th Most Common Venue       10th Most Common Venue  
0           Event Space         Ethiopian Restaurant  
1    Falafel Restaurant                  Event Space  
2                  Farm           Falafel Restaurant  
3        Farmers Market                         Farm  
4  Ethiopian Restaurant  Eastern European Restaurant

In [304]:
toronto_merged.dropna(subset=["Cluster Labels"],inplace=True)

In [307]:
toronto_merged["Cluster Labels"]=toronto_merged["Cluster Labels"].astype(int)

Finally, let's visualize the resulting clusters

In [308]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Now we will Examine Clusters to see what distinguises them.

Cluster 1

In [309]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14        North York               0            Playground   
40  Downtown Toronto               0                  Park   
44  Downtown Toronto               0                  Park   
48        North York               0                  Park   
50        North York               0                  Park   
64         Etobicoke               0                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                  Park   Distribution Center   Filipino Restaurant   
40     Convenience Store         Metro Station               Dog Run   
44              Bus Line           Swim School         Women's Store   
48                 Trail            Restaurant         Women's Store   
50            Playground                 Trail        Discount Store   
64                  Park      Sushi Restaurant         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
14  Fast Food Restaurant        Farmers Market                         Farm   
40   Filipino Restaurant  Fast Food Restaurant               Farmers Market   
44      Doner Restaurant   Filipino Restaurant         Fast Food Restaurant   
48   Distribution Center  Fast Food Restaurant               Farmers Market   
50  Fast Food Restaurant        Farmers Market                         Farm   
64         Women's Store            Donut Shop  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14    Falafel Restaurant           Event Space   Ethiopian Restaurant  
40                  Farm    Falafel Restaurant            Event Space  
44        Farmers Market                  Farm     Falafel Restaurant  
48                  Farm    Falafel Restaurant            Event Space  
50    Falafel Restaurant           Event Space   Ethiopian Restaurant  
64     Electronics Store  Ethiopian Restaurant            Event Space

Cluster 2

In [310]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels     1st Most Common Venue  \
0        Scarborough               1      Fast Food Restaurant   
1        Scarborough               1                       Bar   
2        Scarborough               1        Mexican Restaurant   
3        Scarborough               1               Coffee Shop   
4        Scarborough               1       Fried Chicken Joint   
6        Scarborough               1            Discount Store   
7        Scarborough               1                    Bakery   
8        Scarborough               1                     Motel   
9        Scarborough               1                      Farm   
10       Scarborough               1         Indian Restaurant   
11        North York               1            Breakfast Spot   
12        North York               1            Breakfast Spot   
13        North York               1               Pizza Place   
15        North York               1      Fast Food Restaurant   
17        North York               1  Mediterranean Restaurant   
18        North York               1            Clothing Store   
19        North York               1        Chinese Restaurant   
20        North York               1        Chinese Restaurant   
22        North York               1          Ramen Restaurant   
23        North York               1                      Park   
24        North York               1                      Park   
25        North York               1                      Park   
26        North York               1                      Park   
27        North York               1                Restaurant   
28         East York               1                      Bank   
29         East York               1                       Bar   
30      East Toronto               1                   Airport   
31         East York               1             Grocery Store   
32         East York               1            Baseball Field   
34      East Toronto               1               Pizza Place   
35      East Toronto               1               Pizza Place   
36      East Toronto               1                       Spa   
37   Central Toronto               1                     Trail   
38   Central Toronto               1       Sporting Goods Shop   
39  Downtown Toronto               1         Indian Restaurant   
41  Downtown Toronto               1          Greek Restaurant   
42  Downtown Toronto               1      Fast Food Restaurant   
43  Downtown Toronto               1                      Café   
45  Downtown Toronto               1                     Hotel   
46  Downtown Toronto               1            Clothing Store   
47  Downtown Toronto               1               Pizza Place   
49   Central Toronto               1                       Pub   
51        North York               1                Restaurant   
52              York               1               Coffee Shop   
53              York               1               Coffee Shop   
54  Downtown Toronto               1            Clothing Store   
55      West Toronto               1               Coffee Shop   
56      West Toronto               1               Coffee Shop   
57      West Toronto               1               Coffee Shop   
58        North York               1               Coffee Shop   
59              York               1               Coffee Shop   
60              York               1               Coffee Shop   
61  Downtown Toronto               1               Coffee Shop   
62         Etobicoke               1        Italian Restaurant   
65        North York               1                      Café   
66        North York               1                      Café   
67              York               1                      Café   

         2nd Most Common Venue      3rd Most Common Venue  \
0                   Print Shop              Women's Store   
1                Moving Target                    Dog Run   
2

Cluster 3

In [311]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
33  East York               2    Athletics & Sports        Discount Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
33          Liquor Store         Grocery Store         Women's Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
33  Eastern European Restaurant     Electronics Store  Ethiopian Restaurant   

   9th Most Common Venue 10th Most Common Venue  
33           Event Space                   Farm

Cluster 4

In [312]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
63  Etobicoke               3          Home Service                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
63         Women's Store               Dog Run   Filipino Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
63  Fast Food Restaurant        Farmers Market                  Farm   

   9th Most Common Venue 10th Most Common Venue  
63    Falafel Restaurant            Event Space

Cluster 5

In [313]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5  Scarborough               4            Playground   

        2nd Most Common Venue 3rd Most Common Venue  \
5  Construction & Landscaping      Hakka Restaurant   

        4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
5  Financial or Legal Service  Fast Food Restaurant        Farmers Market   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5                  Farm    Falafel Restaurant           Event Space   

  10th Most Common Venue  
5   Ethiopian Restaurant